packages/blaze-tx/src/tx.ts

```diff
@@ -1241,167 +1243,72 @@ export class TxBuilder {
-        let adaAmount = 0n;
-        // Check the provided collateral inputs to see if we can build valid collateral
-        for (
-          let i = 0;
-          i <
-          Math.min(this.params.maxCollateralInputs, providedCollateral.length);
-          i++
-        ) {
-          adaAmount += providedCollateral[i]!.output().amount().coin();
-          collateral.push(providedCollateral[i]!);
-          if (adaAmount >= 5_000_000n) {
-            break;
-          }
-        }
-        // If we still haven't reached the necessary collateral amount, try to use any available UTxO
-        if (adaAmount < 5_000_000n) {
-          // create a sorted list of utxos by ada amount
-          const adaUtxos = [...this.utxos.values()].sort((a, b) => {
-            const aCoinAmount = this.getUtxoEffectiveCoin(a);
-            const bCoinAmount = this.getUtxoEffectiveCoin(b);
-            return aCoinAmount < bCoinAmount ? -1 : 1;
-          });
-          for (
-            let i = 0;
-            i < Math.min(this.params.maxCollateralInputs, adaUtxos.length);
-            i++
-          ) {
-            adaAmount += this.getUtxoEffectiveCoin(adaUtxos[i]!);
-            collateral.push(adaUtxos[i]!);
-            if (adaAmount >= 5_000_000n) {
-              break;
-            }
-          }
-        }
-        if (adaAmount <= 5_000_000) {
-          throw new Error(
-            "prepareCollateral: could not find enough collateral (5 ada minimum)",
-          );
-        }
-        best = collateral;
-      }
-    }
-    // Set the selected UTXO as collateral in the transaction body.
-    const tis = CborSet.fromCore([], TransactionInput.fromCore);
-    tis.setValues(best.map((x) => x.input()));
-    this.body.setCollateral(tis);
+        remainingLovelace -= selectedValue.coin();
+        const currentCollateral = this.body.collateral()?.values() || [];
+        const collateralSet = CborSet.fromCore([], TransactionInput.fromCore);
+        collateralSet.setValues([
+          ...currentCollateral,
+          ...selectedInputs.map((si) => si.input()),
+        ]);
 
-    for (const bestUtxo of best) {
-      const key = bestUtxo.output().address().getProps().paymentPart!;
-      if (key.type == CredentialType.ScriptHash) {
-        this.requiredNativeScripts.add(key.hash);
-      } else {
-        this.requiredWitnesses.add(HashAsPubKeyHex(key.hash));
+        const totalCollateralInput = value.sum(
+          selectedInputs.map((si) => si.output().amount()),
+        );
+        const collateralReturn = value.sub(
+          totalCollateralInput,
+          new Value(requiredLovelace),
+        );
+
+        this.body.setCollateral(collateralSet);
+        this.body.setTotalCollateral(requiredLovelace);
+        this.body.setCollateralReturn(
+          new TransactionOutput(
+            this.collateralChangeAddress ?? this.changeAddress!,
+            collateralReturn,
+          ),
+        );
```

Comment:
Do we need to calculate the new fee here and run it again in case that changed the required collateral?

Reply:
Good call out, this did change the fee (which might be why Blaze was under estimating the fee before).